src/widgets/canvas.rs

```diff
@@ -176,6 +227,106 @@ impl Grid for CharGrid {
     }
 }
 
+/// The HalfBlockGrid is a grid made up of cells each containing a half block character.
+/// This makes it possible to draw shapes with a resolution of 1x2 dots per cell. This is useful
+/// when you want to draw shapes with a higher resolution than a CharGrid but lower than a
+/// BrailleGrid. This grid type supports a foreground and background color for each terminal cell.
+/// This allows for more flexibility than the BrailleGrid which only supports a single foreground
+/// color for each 2x4 dots cell.
+#[derive(Debug, Default, Clone, Eq, PartialEq, Hash)]
+struct HalfBlockGrid {
+    /// width of the grid in number of terminal columns
+    width: u16,
+    /// height of the grid in number of terminal rows
+    height: u16,
+    /// represents a single color for each cell
+    cells: Vec<Vec<Color>>,
+}
+
+impl HalfBlockGrid {
+    /// Create a new HalfBlockGrid with the given width and height measured in terminal columns and
+    /// rows respectively.
+    fn new(width: u16, height: u16) -> HalfBlockGrid {
+        HalfBlockGrid {
+            width,
+            height,
+            cells: vec![vec![Color::Reset; width as usize]; height as usize * 2],
+        }
+    }
+}
+
+impl Grid for HalfBlockGrid {
+    fn width(&self) -> u16 {
+        self.width
+    }
+
+    fn height(&self) -> u16 {
+        self.height
+    }
+
+    fn resolution(&self) -> (f64, f64) {
+        (
+            f64::from(self.width) - 1.0,
+            f64::from(self.height) * 2.0 - 1.0,
+        )
+    }
+
+    fn save(&self) -> Layer {
+        // We use the following 4 states to represent the 4 possible combinations of foreground and
+        // background colors for each cell:
+        // - off: ' ' colors: reset/reset
+        // - top: '▀' colors: upper/reset
+        // - bottom: '▄' colors: lower/reset (use lower half block set the fg to the lower color)
+        // - both: '▀' colors upper/lower
```

Comment:
It took me a minute to understand this comment. I think it's not going to be super clear what is going on here for someone that doesn't understand this technique. Can we expand this comment further? 

Here's my attempt at expanding it:

```suggestion
        // In terminals, text is usually twice as tall as it is wide,
        // Let's assume that the smallest pixel is called a "Cell".
        // In Unicode, there are a couple of vertical half block characters
        // Namely, the upper half block '▀' and lower half block '▄'. 
        // Together with an empty space ' ' and a full block '█', we can
        // effectively double the resolution of a single Cell.
        // Specifically, for the half block characters, we can use 
        // different foreground and background colors to double the 
        // resolution of a single Cell
        //
        // This technique requires scanning the grid and comparing two
        // elements in the vertical axis to decide what character should be
        // printed in a particular cell.
        //
        // In this scan comparison, there are the 4 states we will encounter:
        // - off: ' ' colors - fg: doesn't matter / bg: reset
        // - top: '▀' colors - fg: upper / bg: reset
        // - bottom: '▀' colors - fg: reset / bg: lower
        // - both: '▀' colors - fg: upper / bg: lower
        //
        // However, because the foreground reset color (i.e. default foreground color) is usually not the same as the background reset color (i.e. default background color), we need to swap around the colors for the `bottom` state. So these are the 4 state we use to represent each cell:
        // - off: ' ' colors - fg: doesn't matter / bg: reset
        // - top: '▀' colors - fg: upper / bg: reset
        // - bottom: '▄' colors: fg: lower / bg: reset (use lower half block instead and set the fg to the lower color)
        // - both: '▀' colors - fg: upper / bg: lower
```

Reply:
Sounds good - I'll tighten that up a little and push a change

Reply:
https://github.com/ratatui-org/ratatui/pull/550/commits/30fbec7f5c91f59e5a0a20d8eccb4a3bae5c916b